orderer/consensus/etcdraft/chain.go

```diff
@@ -921,37 +922,42 @@ func (c *Chain) ordered(msg *orderer.SubmitRequest) (batches [][]*common.Envelop
 			}
 		}
 
-		if !atomic.CompareAndSwapUint32(&c.leaderTransferFailed, 1, 0) {
-			if c.checkForEvictionNCertRotation(msg.Payload) {
-				go func() {
-					if transferee, err := c.Node.abdicateLeader(c.raftID); err != nil {
-						if err.Error() == LEADERTRANSFERTIMEOUT.Error() && transferee != raft.None {
-							oldTransferee := atomic.LoadUint64(&c.leadershipTransferCandidate)
-							if !atomic.CompareAndSwapUint64(&c.leadershipTransferCandidate, oldTransferee, transferee) {
-								atomic.StoreUint32(&c.leaderTransferFailed, 1)
-							}
-							// sleep for tick time & reattempt
-							timer := c.Node.clock.NewTimer(time.Duration(c.Node.config.ElectionTick) * c.Node.tickInterval)
-							defer timer.Stop()
-
-							select {
-							case <-timer.C():
-							case <-c.Node.chain.doneC:
-							}
-						} else if err.Error() == LEADERTRNSFERCALLEDBYFOLLOWER.Error() {
-							c.logger.Debugf("Leadership transfer completed")
-						} else {
-							atomic.StoreUint32(&c.leaderTransferFailed, 1)
-						}
+		if c.checkForEvictionNCertRotation(msg.Payload) && atomic.CompareAndSwapUint32(&c.leadershipTransferInProgress, 0, 1) {
+			go func() {
+				if err := c.Node.abdicateLeader(c.raftID); err != nil {
+					// If there is no leader, abort and do not retry.
+
+					// Return early to prevent re-submission of the transaction
+					if err == ErrNoLeader || err == ErrChainHalting {
+						atomic.StoreUint32(&c.consecutiveAbdicationFailures, 0)
+						return
 					}
-					if err := c.Submit(msg, 0); err != nil {
-						c.logger.Errorf("Transaction proposal failed with error: %v", err)
+
+					if atomic.LoadUint32(&c.consecutiveAbdicationFailures) == 2 {
+						c.logger.Warnw("Abdication failed too many times consecutively, aborting retries")
+						// Reset failure counter as we do not retry anymore.
+						atomic.StoreUint32(&c.consecutiveAbdicationFailures, 0)
+						return
 					}
-				}()
-				return nil, false, nil
-			}
-		} else {
-			atomic.StoreUint64(&c.leadershipTransferCandidate, raft.None)
+
+					// If the error isn't any of the below, it's a programming error, so panic.
+					if err != ErrNoAvailableLeaderCandidate && err != ErrNotALeader && err != ErrTimedOutLeaderTransfer {
+						c.logger.Panicf("Programming error, abdicateLeader() returned with an unexpected error: %v", err)
+					}
+
+					// Else, it's one of the errors above, so we increment the consecutive failure counter and retry.
+					atomic.AddUint32(&c.consecutiveAbdicationFailures, 1)
+				} else {
+					atomic.StoreUint32(&c.consecutiveAbdicationFailures, 0)
+				}
+
+				atomic.StoreUint32(&c.leadershipTransferInProgress, 0)
+
+				if err := c.Submit(msg, 0); err != nil {
```

Comment:
instead of following the submit workflow, can we use the simple loop within the go routine to retry abdicateLeader with a threadhold

```
     if c.checkForEvictionNCertRotation(msg.Payload)  {
                if atomic.CompareAndSwapUint32(&c.leadershipTransferInProgress, 0, 1) {
        		go func() {
                                retryCount = 0
        			for {
                                    if err := c.Node.abdicateLeader(c.raftID); err != nil {
                                          retryCount += 1
                                          if err == ErrNoLeader || err == ErrChainHalting || retryCount == 2 {
                                             atomic.StoreUint32(&c.leadershipTransferInProgress, 0)
                                              return
                                          }
                                    } else {
                                          atomic.StoreUint32(&c.leadershipTransferInProgress, 0)
                                          return
                                    }
                               }
                           }
                  } else {
                           // reject the submit request as already config transaction inprogress
                  }
```